File: katas/php/OOP/encapsulation/encapsulation-bad.php
<?php
// Violación de Encapsulamiento: Datos expuestos sin protección
// ❌ Problema: Cualquiera puede modificar el estado sin validación

class BankAccount {
    // ❌ Propiedades públicas permiten acceso directo sin control
    public float $balance;
    public string $accountNumber;
    public bool $isActive;
    public array $transactionHistory;

    public function __construct(string $accountNumber, float $initialBalance) {
        $this->accountNumber = $accountNumber;
        $this->balance = $initialBalance;
        $this->isActive = true;
        $this->transactionHistory = [];
    }

    // ❌ Método sin validaciones
    public function deposit(float $amount): void {
        $this->balance += $amount;
        $this->transactionHistory[] = "Depósito: $amount";
    }

    // ❌ Método sin validaciones
    public function withdraw(float $amount): void {
        $this->balance -= $amount;
        $this->transactionHistory[] = "Retiro: $amount";
    }
}

// ❌ Sin encapsulamiento, cualquiera puede romper las reglas
echo "=== Violación de Encapsulamiento ===" . PHP_EOL;

$account = new BankAccount("001", 1000);
echo "Balance inicial: $" . $account->balance . PHP_EOL;

// ❌ Modificación directa sin validación
$account->balance = -5000; // ¡Balance negativo!
echo "Balance después de modificación directa: $" . $account->balance . PHP_EOL;

// ❌ Cambiar número de cuenta directamente
$account->accountNumber = "999"; // ¡No debería ser posible!
echo "Número de cuenta modificado: " . $account->accountNumber . PHP_EOL;

// ❌ Desactivar cuenta sin proceso adecuado
$account->isActive = false; // Bypass de reglas de negocio
echo "Cuenta activa: " . ($account->isActive ? "true" : "false") . PHP_EOL;

// ❌ Modificar historial directamente
$account->transactionHistory = []; // ¡Borrar evidencia!
echo "Historial manipulado: " . count($account->transactionHistory) . " transacciones" . PHP_EOL;

// ❌ Retiro sin validación de fondos suficientes
$account->balance = 100;
$account->withdraw(1000); // ¡Retiro mayor al balance!
echo "Balance después de retiro excesivo: $" . $account->balance . PHP_EOL;

File: katas/python/OOP/encapsulation/encapsulation_bad.py
# Violación de Encapsulamiento: Datos expuestos sin protección
# ❌ Problema: Cualquiera puede modificar el estado sin validación

from typing import List


class BankAccount:
    # ❌ Propiedades públicas permiten acceso directo sin control
    def __init__(self, account_number: str, initial_balance: float):
        self.account_number = account_number
        self.balance = initial_balance
        self.is_active = True
        self.transaction_history: List[str] = []

    # ❌ Método sin validaciones
    def deposit(self, amount: float) -> None:
        self.balance += amount
        self.transaction_history.append(f"Depósito: {amount}")

    # ❌ Método sin validaciones
    def withdraw(self, amount: float) -> None:
        self.balance -= amount
        self.transaction_history.append(f"Retiro: {amount}")


if __name__ == "__main__":
    # ❌ Sin encapsulamiento, cualquiera puede romper las reglas
    print("=== Violación de Encapsulamiento ===")

    account = BankAccount("001", 1000)
    print(f"Balance inicial: ${account.balance}")

    # ❌ Modificación directa sin validación
    account.balance = -5000  # ¡Balance negativo!
    print(f"Balance después de modificación directa: ${account.balance}")

    # ❌ Cambiar número de cuenta directamente
    account.account_number = "999"  # ¡No debería ser posible!
    print(f"Número de cuenta modificado: {account.account_number}")

    # ❌ Desactivar cuenta sin proceso adecuado
    account.is_active = False  # Bypass de reglas de negocio
    print(f"Cuenta activa: {account.is_active}")

    # ❌ Modificar historial directamente
    account.transaction_history = []  # ¡Borrar evidencia!
    print(
        f"Historial manipulado: {len(account.transaction_history)} transacciones")

    # ❌ Retiro sin validación de fondos suficientes
    account.balance = 100
    account.withdraw(1000)  # ¡Retiro mayor al balance!
    print(f"Balance después de retiro excesivo: ${account.balance}")
